File: apps/beavernest-be/src/BeaverNestBe/Domain/Diagnostics.fs
module BeaverNestBe.Domain.Diagnostics

open System
open BeaverNestBe.Domain.Readiness

/// Safe ready-only data that HTTP can serialize without exposing host, process,
/// database, or exception detail.
type DiagnosticsReadyResponse =
    { Status: string
      Version: string
      UptimeSeconds: int64
      ServerTimeUtc: DateTimeOffset
      Components: ReadinessComponents }

/// Safe unavailable data deliberately omits version, clock, uptime, and cause.
type DiagnosticsUnavailableResponse =
    { Status: string
      Components: ReadinessComponents }

type DiagnosticsResult =
    | DiagnosticsReady of DiagnosticsReadyResponse
    | DiagnosticsUnavailable

let private readyComponents: ReadinessComponents =
    { Database = "ready"
      Schema = "current" }

let unavailableResponse: DiagnosticsUnavailableResponse =
    { Status = "unavailable"
      Components =
        { Database = "unavailable"
          Schema = "unknown" } }

let readyResponse (version: string) (uptime: TimeSpan) (serverTimeUtc: DateTimeOffset) : DiagnosticsReadyResponse =
    { Status = "ready"
      Version = version
      UptimeSeconds = max 0L (int64 (Math.Floor(uptime.TotalSeconds)))
      ServerTimeUtc = serverTimeUtc.ToUniversalTime()
      Components = readyComponents }

File: apps/beavernest-app/lib/domain/diagnostics.dart
import 'package:beavernest_app/domain/readiness.dart';

/// A deliberately small operational snapshot safe to display in the workspace.
final class WorkspaceDiagnostics {
  const WorkspaceDiagnostics.ready({
    required this.version,
    required this.uptimeSeconds,
    required this.serverTimeUtc,
    required this.database,
    required this.schema,
  }) : isAvailable = true;

  const WorkspaceDiagnostics.unavailable({
    required this.database,
    required this.schema,
  }) : isAvailable = false,
       version = null,
       uptimeSeconds = null,
       serverTimeUtc = null;

  final bool isAvailable;
  final String? version;
  final int? uptimeSeconds;
  final String? serverTimeUtc;
  final DatabaseReadiness database;
  final SchemaReadiness schema;
}
